doc/specs/#885 - winrt Terminal Settings.md

```diff
@@ -0,0 +1,133 @@
+---
+author: Carlos Zamora @carlos-zamora
+created on: 2020-07-10
+last updated: 2020-07-10
+issue id: [#885]
+---
+
+# Spec Title
+
+## Abstract
+
+This spec proposes a major refactor and repurposing of the TerminalSettings project. TerminalSettings would
+ be responsible for exposing, serializing, and deserializing settings as WinRT objects for Windows Terminal. In
+ doing so, Terminal's settings model is accessible as WinRT objects to existing components like TerminalApp,
+ TerminalControl, and TerminalCore. Additionally, Terminal Settings can be used by the Settings UI or
+ Shell Extensions to modify or reference Terminal's settings respectively.
+
+## Inspiration
+
+The main driver for this change is the Settings UI. The Settings UI will need to read and modify Terminal Settings.
+ At the time of writing this spec, the Terminal Settings are serialized as objects in the TerminalApp project.
+ To access these objects, the Settings UI would need to be a part of TerminalApp too, making it more bloated.
+
+## Solution Design
+
+### Terminal Settings Model: Objects and Projections
+
+The objects/interfaces introduced in this section will be exposed as WinRT objects/interfaces
+ within the `Microsoft.Terminal.Settings` namespace. This allows them to be interacted with across
+ different project layers.
+
+As before, a terminal instance will need a `Profile` (formerly `TerminalSettings`) to be able to update its settings.

```

Comment:
So, okay, we need to move ICore and IControl down to the Core and Control layers and have a bridge in TerminalApp that connects a tuple `{Settings, Active Profile}` up as an `ICore+IControl` combo. Profile doesn't need to implement those interfaces itself.

Reply:
We *need* to find a way to delete the entire TerminalSettings project. We're holding COM wrong - all we have to do is implement the right interfaces, we only have an extra dll and runtimeclass because it was annoying to do otherwise with C++/WinRT.

From where I sit, it looks like..

![image](https://user-images.githubusercontent.com/189190/87474780-84369980-c5d8-11ea-92a7-58add01163ee.png)

Reply:
In an ideal world, yea, that's how it _should_ be, but the TerminalSettings project exists because we couldn't get the TerminalCore project to emit a winmd for ICoreSettings without implementing at least one type, which it _doesn't_ 😑

Reply:
But like, this is our chance to uncut the corners we cut in the leadup to 0.1. I can't countenance speccing a redesign when that spec says "continue doing the bad thing" :smile:

Making idl work to product interfaces is some dark MIDL2 magic, but I think that it's possible. There's actually an example now in the DX code (IBoxDrawingEffect) that does it!

Reply:
If we can do that, I'll ACK the change. But definitely with the `AppSettings(Profile) -> IControlSettings&ICoreSettings` change